### PR TITLE
Issue/1767 download status

### LINKED ISSUE #1767: Add support for checking the status of a backup preparation download
The goal of this ticket is to add support for checking the status of a backup preparation download

GET https://public-api.wordpress.com/wpcom/v2/sites/182675569/rewind/downloads

The response will take the following form:

RUNNING
```
{
  "downloadId": 276158,
  "rewindId": "1603788401.423",
  "backupPoint": "2020-10-27T08:46:41+00:00",
  "startedAt": "2020-10-27T11:00:23+00:00",
  "progress": 49
}
```

FINISHED

```
{
  "downloadId": 276158,
  "rewindId": "1603788401.423",
  "backupPoint": "2020-10-27T08:46:41+00:00",
  "startedAt": "2020-10-27T11:00:23+00:00",
  "downloadCount": 0,
  "validUntil": "2020-10-28T11:00:25+00:00",
  "url": "https://public-api.wordpress.com/wpcom/v2/sites/182675569/rewind/downloads/276158/data?token=6VxkPOSkvwcLQfhPRMIh9TCNaJKygYueLjnzzS5wNl1R05dnc3jKEvmgAuCbmhLp"
}

```

### PR DESCRIPTION
**Pull request by zwarm**
Fixes #1767 

This PR adds support for checking download preparation status.
GET https://public-api.wordpress.com/wpcom/v2/sites/182675569/rewind/downloads

Note: I named everything "download", but after submitting this PR I decided to rename methods/objects to BackupDownload so it has more context. That will be in a separate PR following approval of this one.

This PR follows along with the same patterns used for checking rewind status.

P.S. I also adjusted minor Kotlin inspection finds such as redundant qualifiers.

**To Test**
The feature has not been implemented yet. CI tests passing are okay for this PR.

### CODE REVIEW

File: example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt [hunk 1]
```diff
@@ -395,6 +397,51 @@ class ActivityLogRestClientTest {
...
+    @Test
+    fun fetchActivityDownload_dispatchesGenericErrorOnFailure() = test {
+        initFetchDownloadStatus(error = WPComGsonNetworkError(BaseNetworkError(NETWORK_ERROR)))
+
+        val payload = activityRestClient.fetchActivityDownload(site)
+
+        assertEmittedDownloadStatusError(payload, DownloadStatusErrorType.GENERIC_ERROR)
+    }
+
+    private fun assertEmittedDownloadStatusError(
```
Comment (ParaskP7) [inline example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt:409]: Question: Why did this code extracted to a private function? Will it be reused?

File: example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt [hunk 2]
```diff
@@ -395,6 +397,51 @@ class ActivityLogRestClientTest {
...
+        payload: FetchedDownloadStatePayload,
+        errorType: DownloadStatusErrorType
+    ) {
+        with(payload) {
+            assertEquals(this.site, site)
+            assertTrue(this.isError)
+            assertEquals(errorType, this.error.type)
```
Comment (ParaskP7) [inline example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt]: Minor: The `this` keyword is not needed here, is it?

File: example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt [hunk 3]
```diff
@@ -395,6 +397,51 @@ class ActivityLogRestClientTest {
...
+        }
+    }
+
+    @Test
+    fun fetchActivityDownload_dispatchesResponseOnSuccess() = test {
```
Comment (ParaskP7) [inline example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt:421]: Minor: I have a few suggestion for improving this test, instead writing all those down for you, how about if I provide the code snippet?
```
    @Test
    fun fetchActivityDownload_dispatchesResponseOnSuccess() = test {
        val downloadResponse = DOWNLOAD_STATUS_RESPONSE.copy(progress = 55)
        initFetchDownloadStatus(downloadResponse)

        val payload = activityRestClient.fetchActivityDownload(site)

        with(payload) {
            assertEquals(site, site)
            assertNull(error)
            assertNotNull(downloadStatusModelResponse)
            downloadStatusModelResponse?.apply {
                assertEquals(downloadId, downloadResponse.downloadId)
                assertEquals(rewindId, downloadResponse.rewindId)
                assertEquals(backupPoint, downloadResponse.backupPoint)
                assertEquals(startedAt, downloadResponse.startedAt)
                assertEquals(downloadCount, downloadResponse.downloadCount)
                assertEquals(validUntil, downloadResponse.validUntil)
                assertEquals(url, downloadResponse.url)
                assertEquals(progress, downloadResponse.progress)
            }
        }
    }
```

Wdyt Annmarie?

Question: For future reference, would you like me to continue doing so or would you instead prefer it written no matter how long?

File: example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt [hunk 4]
```diff
@@ -365,6 +368,47 @@ class ActivityLogStoreTest {
...
+                )
+        )
+        val action = ActivityLogActionBuilder.newFetchDownloadStateAction(payload)
+        activityLogStore.onAction(action)
+
+        verify(activityLogRestClient).fetchActivityDownload(siteModel)
+    }
+
+    @Test
+    fun storeFetchedDownloadStatusToDb() = test {
```
Comment (ParaskP7) [inline example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt:387]: Question: How about testing the error path as well?

File: example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt [hunk 5]
```diff
@@ -365,6 +368,47 @@ class ActivityLogStoreTest {
...
+        val downloadStatusModel = mock<DownloadStatusModel>()
+        val payload = FetchedDownloadStatePayload(downloadStatusModel, siteModel)
+        whenever(activityLogRestClient.fetchActivityDownload(siteModel)).thenReturn(payload)
+
+        val fetchAction = ActivityLogActionBuilder.newFetchDownloadStateAction(FetchDownloadStatePayload(siteModel))
+        activityLogStore.onAction(fetchAction)
+
+        verify(activityLogSqlUtils).replaceDownloadStatus(siteModel, downloadStatusModel)
+        val expectedChangeEvent = ActivityLogStore.OnDownloadStatusFetched(ActivityLogAction.FETCH_DOWNLOAD_STATE)
+        verify(dispatcher).emitChange(eq(expectedChangeEvent))
```
Comment (ParaskP7) [inline example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt:397]: Question: Can this be considered as part of this db test? It seems that it can be a separate test altogether?

File: example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt [hunk 6]
```diff
@@ -365,6 +368,47 @@ class ActivityLogStoreTest {
...
+    @Test
+    fun returnDownloadStatusFromDb() {
+        val downloadStatusModel = mock<DownloadStatusModel>()
+        whenever(activityLogSqlUtils.getDownloadStatusForSite(siteModel))
+                .thenReturn(downloadStatusModel)
+
+        val downloadStatusFromDb = activityLogStore.getDownloadStatusForSite(siteModel)
+
+        verify(activityLogSqlUtils).getDownloadStatusForSite(siteModel)
+        assertEquals(downloadStatusModel, downloadStatusFromDb)
```
Comment (ParaskP7) [inline example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt]: Question: Does this assertion provide actual value for this test? How does it help?

File: fluxc/src/main/java/org/wordpress/android/fluxc/model/activity/DownloadStatusModel.kt [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package org.wordpress.android.fluxc.model.activity
+
+import java.util.Date
+
+data class DownloadStatusModel(
```
Comment (ParaskP7) [inline fluxc/src/main/java/org/wordpress/android/fluxc/model/activity/DownloadStatusModel.kt:5]: Idea: I did leave the very same comment on @ashiagr's PR about domain modelling. I am just going to comment it again so we can spin up a conversation on that again (if that's ok).

My suggestion is that this model can be split into two separate models, using a sealed class. Something like a `DownloadStatusRunningModel` and a `DownloadStatusFinishedModel` model. This way there would be no nullable values per model and the usage of those later on would be much more straightforward. Wdyt?

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/ActivityLogSqlUtils.kt [hunk 8]
```diff
@@ -170,6 +196,21 @@ class ActivityLogSqlUtils
...
+                localSiteId = site.id,
+                remoteSiteId = site.siteId,
+                downloadId = this.downloadId,
+                rewindId = this.rewindId,
+                backupPoint = this.backupPoint.time,
+                startedAt = this.startedAt.time,
+                progress = this.progress,
+                downloadCount = this.downloadCount,
+                validUntil = this.validUntil?.time,
+                url = this.url
```
Comment (ParaskP7) [inline fluxc/src/main/java/org/wordpress/android/fluxc/persistence/ActivityLogSqlUtils.kt:210]: Minor: The `this` keyword is not needed here, is it?

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/ActivityLogSqlUtils.kt [hunk 9]
```diff
@@ -289,4 +330,43 @@ class ActivityLogSqlUtils
...
+                    downloadId,
+                    rewindId,
+                    Date(backupPoint),
+                    Date(startedAt),
+                    progress,
+                    downloadCount,
+                    validUntil?.let {
+                        Date(it)
+                    },
+                    url
```
Comment (ParaskP7) [inline fluxc/src/main/java/org/wordpress/android/fluxc/persistence/ActivityLogSqlUtils.kt:368]: Idea: Maybe the usage of named arguments could help here to verify correctness without the need to jump to and back into their declaration?

**Review by ParaskP7 — COMMENTED** [review]
Thanks again for including me in this PR Annmarie! 🌟 

As always, I have only questions, minors and ideas as comments, still learning in progress, please bare with me! 📚 

Also, I have a couple of extra generic things that I want to ask:
1) The `ActivityFixture` file can be renamed file to include the `Log` word in it. Actually a lot of names are missing the `Log` word in them. Maybe someone can spend some time and create a separate PR to improve on that. To me, having it as plain `Activity` can be confusing. What is even worse, having some of the class/file names as `Activity`, while others as `ActivityLog` can create even more confusion in the future. Also, the package name might need to change to include 'log' as well. Wdyt?
1) I always suggest and trying hard myself to separate the refactoring from the feature commits in order to make my coding and later on the code review more targeted and easy. How do you all handling that, do you prefer one over the other?

**Comment by zwarm** [comment]
> 1. The `ActivityFixture` file can be renamed file to include the `Log` word in it. Actually a lot of names are missing the `Log` word in them. Maybe someone can spend some time and create a separate PR to improve on that. To me, having it as plain `Activity` can be confusing. What is even worse, having some of the class/file names as `Activity`, while others as `ActivityLog` can create even more confusion in the future. Also, the package name might need to change to include 'log' as well. Wdyt?

This is a good suggestion, please go ahead and create a low priority issue for this. 

> 2. I always suggest and trying hard myself to separate the refactoring from the feature commits in order to make my coding and later on the code review more targeted and easy. How do you all handling that, do you prefer one over the other?

I don't have a preference, so whatever works for you.

File: example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt (hunk 1 above)
Comment (zwarm) [reply example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt:409]: It very well may be. This is the generic error and until I get to the UI implementation, I don't know if there will be additional error situations I will need to handle.

File: example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt (hunk 5 above)
Comment (zwarm) [reply example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt:397]: Following existing patterns for rewind

File: example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt (hunk 3 above)
Comment (zwarm) [reply example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt:421]: > Question: For future reference, would you like me to continue doing so or would you instead prefer it written no matter how long?

Thanks for the input, I appreciate the time you took to write this out, but pease don't spend your time rewriting unless there is something horribly wrong. Perhaps we can start sharing these types of things in a monthly learning meetup - kind of like a "writing test best practices" series. Wdyt?

File: example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt (hunk 4 above)
Comment (zwarm) [reply example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt:387]: > Question: How about testing the error path as well? 

All the error paths are unknown at the moment. This is on my future to-do list. thanks

File: example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt (hunk 6 above)
Comment (zwarm) [reply example/src/test/java/org/wordpress/android/fluxc/store/ActivityLogStoreTest.kt]: Removed in latest commit

File: fluxc/src/main/java/org/wordpress/android/fluxc/model/activity/DownloadStatusModel.kt (hunk 7 above)
Comment (zwarm) [reply fluxc/src/main/java/org/wordpress/android/fluxc/model/activity/DownloadStatusModel.kt:5]: > My suggestion is that this model can be split into two separate models, using a sealed class

I really like this idea. Using sealed classes with a stricter structure and not mucking around with nulls is a + in my book.  My only hesitation is that this functionality mimics the existing rewind functionality and I wouldn't want to change one without the other. I want to check with @malinajirka / @ashiagr and see if they are planning to switch out to existing functionality, as I want to stay consistent.

Thanks 👍

File: example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt (hunk 3 above)
Comment (ParaskP7) [reply example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt:421]: Great, yes I agree! Although, I usually will spend that time anyway, it is just in this case I was trying to understand this test and some of the inputs/outputs. I find it helps me playing around so please don't worry about me. Also, I am doing my code review from the IDE, so changing and testing code is a matter of seconds for me, at least in this case.

About the monthly learning meetup and various series such as this, I would totally ❤️  if we can arrange something like that! @malinajirka, @ashiagr is this possible, would you like that as well? Or @zwarm did you mean more of a whole `Android` thingy meetup?

File: fluxc/src/main/java/org/wordpress/android/fluxc/model/activity/DownloadStatusModel.kt (hunk 7 above)
Comment (ParaskP7) [reply fluxc/src/main/java/org/wordpress/android/fluxc/model/activity/DownloadStatusModel.kt:5]: Thanks for the detailed explanation Annmarie! Yeap, please talk it through with the rest and decide on the way you want to go forward, it is just an idea, and consistency is the key!

**Comment by ParaskP7** [comment]
> This is a good suggestion, please go ahead and create a low priority issue for this.

Perfect, will do that then, thanks Annmarie! When I do create the issue (probably tomorrow) I'll link it in this PR for our reference.

File: example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt (hunk 3 above)
Comment (malinajirka) [reply example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt:421]: We had a couple of learning meetups. There was even one about unit test best practices. Unfortunately, it was in-person learnup so there is no recording 😢. I'd love to have more of those!

File: fluxc/src/main/java/org/wordpress/android/fluxc/model/activity/DownloadStatusModel.kt (hunk 7 above)
Comment (malinajirka) [reply fluxc/src/main/java/org/wordpress/android/fluxc/model/activity/DownloadStatusModel.kt:5]: As I mentioned on [the "original" convo](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1763#discussion_r522036285) I really like the sealed classes approach. I also commented there that I wouldn't change `ScanStateModel` without changing `RewindStatusModel`.

However, in this particular case I don't feel strongly about the consistency. "DownloadStatusModel" is not similar to "RewindStatusModel" anyway. The Rewind model contains "status" field so it's at least readable what status it is in. However, the download model doesn't contain any such info so it's very difficult to see that it can actually be in two different states.

So to summarize, I personally wouldn't mind even if we refactored just this class. Having said that, feel free to keep it as is. Whatever you think is the best, tbh I'm not sure.

File: example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt (hunk 1 above)
Comment (malinajirka) [inline example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt:409]: Nitpick: We might want to consider moving all the private methods to the bottom of the class. Wdyt?

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/ActivityLogStore.kt [hunk 10]
```diff
@@ -112,6 +123,13 @@ class ActivityLogStore
         return emitDownloadResult(payload, DOWNLOAD)
     }
 
+    suspend fun fetchActivitiesDownload(
```
Comment (malinajirka) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/ActivityLogStore.kt:126]: Note: I find this name a bit confusing, I'm not sure I understand why the method name includes "activities". I'd have called it something like "fetchDownloadState". But I understand you kept consistency with "fetchActivitiesRewind". Am I missing something or do you find the name confusing too? (no need to change anything, I was just curious if I'm missing something)

**Review by malinajirka — APPROVED** [review]
Great job @zwarm! The code LGTM. I've left just one super minor comment so I'm merging this PR.

**Comment by malinajirka** [comment]
> The ActivityFixture file can be renamed file to include the Log word in it. Actually a lot of names are missing the Log word in them. Maybe someone can spend some time and create a separate PR to improve on that. To me, having it as plain Activity can be confusing. What is even worse, having some of the class/file names as Activity, while others as ActivityLog can create even more confusion in the future. Also, the package name might need to change to include 'log' as well. Wdyt?

@ParaskP7 I agree that the current state can be confusing sometimes. Here's my understanding of the current state. `ActivityLog` refers to the ActivityLog feature (list of events on a site), whereas `Activity` refers to a single item (event). I agree that some classes/files/fields are not following this "rule", so these should be definitely renamed. I'm not sure about the ones which are intentionally called "Activity(ies)", perhaps we could come up with a better name for them (eg. ActivityLogEvent/ActivityLogItem/...). Let's continue with this discussion on the ticket.

**Comment by ParaskP7** [comment]
> @ParaskP7 I agree that the current state can be confusing sometimes. Here's my understanding of the current state. ActivityLog refers to the ActivityLog feature (list of events on a site), whereas Activity refers to a single item (event). I agree that some classes/files/fields are not following this "rule", so these should be definitely renamed. I'm not sure about the ones which are intentionally called "Activity(ies)", perhaps we could come up with a better name for them (eg. ActivityLogEvent/ActivityLogItem/...). Let's continue with this discussion on the ticket.

Thanks for your input @malinajirka! Didn't think about the `ActivityLog` plural and `Activity` as single, it makes sense, yes. Indeed, we should make sure that everything is following this "rule" wherever possible. Still something troubles in me within when reading plain `Activity`, probably I am too used to this keyword being referred to the view layer I guess. Yes, let continue this discussion on the ticket. 🙏

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/ActivityLogStore.kt (hunk 10 above)
Comment (zwarm) [reply fluxc/src/main/java/org/wordpress/android/fluxc/store/ActivityLogStore.kt:126]: Oh yes, I hate the name too and was just staying consistent. I am going to add this change to the rename PR (if it hasn't been closed yet). 🙇

File: example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt (hunk 1 above)
Comment (zwarm) [reply example/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/activity/ActivityLogRestClientTest.kt:409]: I'll add this into my rename PR too! Thanks

**Comment by ParaskP7** [comment]
> Perfect, will do that then, thanks Annmarie! When I do create the issue (probably tomorrow) I'll link it in this PR for our reference.

As suggested, I created this [issue](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/1775) for our future reference.
